### DISCUSSION TITLE
How to produce multi channel output

### DISCUSSION THREAD

**Discussion by RuoyanMeng** [Project-MONAI/MONAILabel#1746] [Q&A]
Hi, I am working on liver segmentation (liver, tumor and vessel), and tried to use monai label and 3D Slicer as a tool for annotation and active learning. I introduced my own model following the [instruction](https://github.com/Project-MONAI/MONAILabel/tree/e31eaee03b4a3e480532de0e01a1d96feef799e1/sample-apps#creating-a-custom-app). 
However, my model predicts 3 classes, and the resulting labels overlap with each other, so they need to be stored separately. For now, I managed to run inference successfully with one label extracted. I wonder if there is a way to return a 3-channel image with all 3 labels non-overlapped. Thank you so much!

**Comment by lassoan** [comment]
Yes, sure. NRRD - the file format that 3D Slicer uses for storing segmentation supports overlapping labels. If you save the results in this format then Slicer will load them correctly. You can also specify segment names, colors, standard terminology, etc. See specification of the file format here: https://slicer.readthedocs.io/en/latest/developer_guide/modules/segmentations.html#segmentation-labelmap-file-format-seg-nrrd

You can use [slicerio](https://pypi.org/project/slicerio/) Python package to conveniently read/write NRRD files with additional segmentation metadata.

**Comment by diazandr3s** [comment]
Hi @RuoyanMeng,

As @lassoan suggested, you can use the NRRD file format to store the prediction.

To do that, you may need to add a custom post-transform [here](https://github.com/Project-MONAI/MONAILabel/blob/e31eaee03b4a3e480532de0e01a1d96feef799e1/sample-apps/radiology/lib/infers/segmentation.py#L91).

Creating a custom transform isn't difficult. Here are some examples: https://github.com/Project-MONAI/MONAILabel/blob/e31eaee03b4a3e480532de0e01a1d96feef799e1/sample-apps/radiology/lib/transforms/transforms.py

Hope this helps,

**Reply by RuoyanMeng** [reply]
Hi @lassoan  and @diazandr3s ,

Thank you for all your suggestions! 

I checked the log info and source code of MONAI-Label, and found that [basic_infer](https://github.com/Project-MONAI/MONAILabel/blob/e31eaee03b4a3e480532de0e01a1d96feef799e1/monailabel/tasks/infer/basic_infer.py#L104C20-L104C36) provide a key to assign the result extension by: `data["result_extension"] = ".nrrd"`. I didn't set this before but when running the MONAI-Label with 3D Slicer, it already used ".nrrd" as an extension. The MONAI-Label nrrd writer works as expected, however, only the first channel "Liver" is shown when the result is visualized in 3d Slicer.

Then I looked into the 3D Slicer plugin [code](https://github.com/Project-MONAI/MONAILabel/blob/4736d5fab7a33edd264a30641759e6aa9b0d8153/plugins/slicer/MONAILabel/MONAILabel.py#L2319) and found that "result_extension" has been set as .nrrd already, and python log from 3D slicer shown a VTK error `[VTK] Input has 3 instead of 1 scalar component`. Hence I dug into [updateSegmentationMask](https://github.com/Project-MONAI/MONAILabel/blob/4736d5fab7a33edd264a30641759e6aa9b0d8153/plugins/slicer/MONAILabel/MONAILabel.py#L1710) function where these [lines](https://github.com/Project-MONAI/MONAILabel/blob/4736d5fab7a33edd264a30641759e6aa9b0d8153/plugins/slicer/MONAILabel/MONAILabel.py#L1726-L1727) showed the possibilities of reading multi-channel segmentation file correctly just by deleting `and self.file_ext == ".seg.nrrd"` and set `data["result_extension"] = ".seg.nrrd"`. And it works, I can finally have all 3 segmentations shown correctly in 3D Slicer!

However, when saving edited labels, a nii.gz file with labels non-overlap with each other will be saved. To solve that, I need set  `self.file_extion, "MONAILabel/fileExtension", fileExtension` with the correct value ".seg.nrrd".  Even though I set `fileExtension.setText(".seg.nrrd")`, the `"MONAILabel/fileExtension"` value will turn to ".nii.gz" and make `self.file_extion` into ".nii.gz".

Anyidea how can I mange the value of `"MONAILabel/fileExtension"` to save the overlapping labels.
Thanks again for your input!

**Reply by diazandr3s** [reply]
Hi @RuoyanMeng,

Have you considered using this transform to save the result in a local folder at the point you wish? https://docs.monai.io/en/stable/transforms.html#saveimaged

@lassoan could you please remind us how 3DSlicer manages multi-tensor images stored in a single file?

**Comment by shuijing1987** [comment]
can I execute the radiology model with some params to acchive this goal ? make the different mask as different channels?

**Reply by shuijing1987** [reply]
> different

I means use the `batch_infer` in command line to generate the `.nrrd` results
